src/protect-function-app/greeting.cs

```diff
@@ -0,0 +1,50 @@
+/*
+This is an Azure Function that responds at GET /api/greeting.
+
+Using the built-in authentication and authorization capabilities (sometimes
+referred to as "Easy Auth") of Azure Functions, offloads part of the authentication
+and authorization process by ensuring that every request to this Azure Function has
+an access token. That access token has had its signature, issuer (iss), expiry
+dates (exp, nbf), and audience (aud) validated. This means all that is left to
+perform is any per-function authorization related to your application.
+*/
+
+using System;
+using System.IO;
+using System.Security.Claims;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Azure.WebJobs;
+using Microsoft.Azure.WebJobs.Extensions.Http;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Api
+{
+    public static class Greeting
+    {
+        /*
+        Because Easy Auth has already validated the signature, the validation is not
+        performed again, but instead the token is is being decoded only to get access
+        to its contained scopes claim.
+        */
+        [FunctionName("greeting")]
+        public static IActionResult Run(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "get")] HttpRequest req,
+            ClaimsPrincipal principal,
+            ILogger log)
+        {
+            // This API endpoint requires the "Greeting.Read" scope to be present, if it is
+            // not, then reject the request with a 403.
+            if (!principal.Claims.Any(c => c.Type == "http://schemas.microsoft.com/identity/claims/scope" && c.Value.Contains("Greeting.Read")))

```

Comment:
Any thought of something more like this?  (Didn't test, but the docs say there is a `HasClaim` method that seems to do this.  And teaching `scp` might be a good side effect here, since that's what Azure AD returns for scopes.

```suggestion
            if (!principal.HasClaim("scp", "Greeting.Read"))
```

Reply:
thanks for the suggestion @ckittel, it looks more elegant now!

done | addressed from [dd9b7](https://github.com/Azure-Samples/ms-identity-docs-code-dotnet/pull/19/commits/dd9b7887693924445457efef1b8bd21c1051bb95)

:warning: major drawback: more than scope as it was tested before and this will return `403`.

Reply:
Oh, so you're saying that if this token has `Greeting.Read Greeting.Write` this will fail now.  That's a big deal, thanks for testing that, and I totally failed to consider that!  Eeek.

Maybe it's best then like `principal.Claims["scp"].Value.Split(' ').Contains("Greeting.Read")`!

(I don't remember what the indexer is on `Claims` -- if it's `[]` or something else. -- so that's just pseudo code above. :)

Reply:
Talked on teams, the indexer approach would require additional error handling to still result in a `403` on a missing required scope.  Back to Linq + exact scope validation (using a split on ` `)